pkg/component/controller/cplb_reconciler.go

```diff
@@ -0,0 +1,175 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// CPLBReconciler monitors the endpoints of the "kubernetes" service in the
+// "default" namespace. It notifies changes though the updateCh channel provided
+// in the constructor.
+type CPLBReconciler struct {
+	log             *logrus.Entry
+	kubeconfigPath  string
+	clientset       kubernetes.Interface
+	addresses       []string
+	mu              sync.RWMutex
+	updateCh        chan<- bool
+	watchCancelFunc context.CancelFunc
+}
+
+func NewCPLBReconciler(kubeconfigPath string, updateCh chan<- bool) *CPLBReconciler {
+	return &CPLBReconciler{
+		log:            logrus.WithField("component", "cplb-reconciler"),
+		kubeconfigPath: kubeconfigPath,
+		updateCh:       updateCh,
+	}
+}
+func (r *CPLBReconciler) getClientSet() error {
+	clientset, err := kubeutil.NewClientFromFile(r.kubeconfigPath)
+	if err != nil {
+		return err
+	}
+
+	r.clientset = clientset
+	return nil
+}
+
+func (r *CPLBReconciler) Start() error {
+	if err := r.getClientSet(); err != nil {
+		return fmt.Errorf("Failed to get clientset. Unable to : %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r.watchCancelFunc = cancel
+	go r.watchAPIServers(ctx)
+
+	return nil
+}
+
+func (r *CPLBReconciler) Stop() {
+	r.log.Info("Stopping CPLB reconciler")
+	r.watchCancelFunc()
+}
+
+func (r *CPLBReconciler) watchAPIServers(ctx context.Context) {
+	// Before starting check that the API is actually responding
+	for {
+		// watch the "kubernetes" endpoints in the "default" namespace
+		watcher, err := r.clientset.CoreV1().Endpoints("default").Watch(ctx, metav1.ListOptions{
+			FieldSelector: "metadata.name=kubernetes",
+		})
+		if err != nil {
+			r.log.Errorf("Failed to watch kubernetes endpoints. Retrying in 5 seconds : %v", err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		r.log.Info("Kubernetes endpoint watcher started")
+	loop:
+		for {
+			select {
+			case event, ok := <-watcher.ResultChan():
+				if !ok {
+					r.log.Warn("Kubernetes endpoint watcher closed, retrying to open it")
+					time.Sleep(5 * time.Second)
+					break loop
+				}
+				switch event.Type {
+				case watch.Added, watch.Modified:
+					endpoint, ok := event.Object.(*corev1.Endpoints)
+					if ok {
+						r.log.Info("Kubernetes endpoint was updated")
+						r.maybeUpdateIPs(endpoint)
+					}
+				case watch.Deleted:
+					r.log.Error("Kubernetes endpoint was deleted, this is not supposed to happen. Ignoring the event")
+				case watch.Error:
+					r.log.Errorf("Kubernetes endpoint watch error: %v", event.Object)
+					time.Sleep(5 * time.Second)
+					break loop
+				}
+			case <-ctx.Done():
+				watcher.Stop()
+				r.log.Info("Stopped watching kubernetes endpoints")
+				return
+			}
+		}
+	}
+}
+
+// maybeUpdateIPs updates the list of IP addresses if the new list has
+// different addresses
+func (r *CPLBReconciler) maybeUpdateIPs(endpoint *corev1.Endpoints) {
```

Comment:
Do we really need to compare the lists and store the "state" internally? What if we'd do just a signal whenever there's changes seen from watch and then always reload all the addreses from the endpoint object itself? I mean they are not changing very often in general and thus keeping a cache kinda feels bit wonky.

Reply:
I had already implemented this in a slightly different way but it's pretty much the same. If you feel strongly about this instead of what I've done let me know and I'll replace it.

Reply:
No, all good!